[docs]: change ideas from md to structure dir tree (#4)
NixOS/GSoC2024: ideas/README.md

# Ideas

## Intro

> Welcome to the Google Summer of Code ideas repository!
> Here you can propose new ideas to participate to this fantastic event 
> and contribute to the nix/nixOS/nixpkgs community ecosystem!

Reviewer: ```suggestion
Welcome to the NixOS Google Summer of Code ideas list!
Here you can propose new ideas to participate in this fantastic event 
and contribute to the NixOS community ecosystem.
```

Below you will find all instructions to propose your ideas, so don't be shy
to contribute!

Summerily yours,
the repo maintainers.

## Submission fields explanation

### ID
The ID is composed by the current year and an incremental 4 digit integer
assigned by one of the repo maintainers and will also be the directory name

> that will contain all the artifacts for that idea.

Author: I think it's just simpler to use numerical IDs instead of having to remember the exact title.

Maybe we can use the PR number associated with the idea as the ID?

Not a huge deal if they think it's not useful, could be my personal preference.

### Title
The title should be clear, concise and inspirative.

It should also be prefixed by a categorey chosen among:
    * nix
    * nixpkgs
    * nixos
Depending on which area of the ecosystem more closely fits your idea.

> ### Author
> Just the author's github handle.
> 
> ### Proposed Menthor
> If you're in contact with a member of the nix community who's interested in
> helping you as your direct menthor, you can mention the github handle here so
> that the repo maintainers wil be added to the conversation in the PR.

Reviewer: ```suggestion
### Possible Menthors

- [@your-self](https://github.com/ghost)

And if you know someone else who might be interested, ask them beforehand and if they are fine with it add them to the list and ping them in the PR adding the idea.
```

### Effort
This is the "size" of the effort level of the idea. It should be one of:
    * SMALL
    * MEDIUM
    * LARGE

> ### Difficulty
> This is how "hard" it is at a technical level to implement the idea. It can be:
>     * EASY
>     * MEDIUM
>     * HARD

Reviewer: ```suggestion
### Difficulty
This is how "hard" it is at a technical level to implement the idea. It can be:
 
- EASY
- MEDIUM
- HARD
```
same here with sticking to markdown conventions

> ### Overview/Abstract
> This is a short description of the idea. Try to stay within the limit of 400
> characters. Think of this as your "elevator pitch".
> ### Description
> Here you can add all the nitty gritty details of the idea. Feel free to add
> anything that you think is relevant for the staff to understand and evaluate
> your idea.

Reviewer: ```suggestion

### Description
Here you should explain what the idea is, and why it would be interesting to work on it.
It should only be 2-5 sentences.
```
lets stick to the way google recommends to write the ideas.

### Linkography
Just the link to the relevant documentation and/or description you wrote
above.

## Idea Template

This would be a PR aimed at creating a README.md file with the content below

> in the directory [ideas/2024/0000](./ideas/2024/0000).

Reviewer: Putting every idea in it's own directory will make it hard to read through all of them at once because you will have to open different pages for each one.
It should be a ideas-*list* so I think we should just put them in one file and sort them by size.
Please remember that this should be a low level entry point with ideas for newbies.
It should also be easy and minimal effort for mentors to propose ideas, we have the [RFC](https://github.com/nixos/rfcs/) process for complex and large proposals.

Author: yes but the RFC process is not related t other GSoC initiative.

I agree about the idea list and that's why I wanted to keep the README.md as a "summary"/"index" which then points to the different directories that can contain all the needed data, including additional files, images, source code files, etc..

Maybe I'm over-optimizing :)

### Metadata

**ID**: [2024/0000](./ideas/2024/0000)

**Title**: [nixpkgs] My idea to nixify the world!

**Author**: @your-github-handle-here

**Proposed Menthor**: [optional] @your-menthor-github-handle-here

**Effort**: SMALL/MEDIUM/LARGE

**Difficulty**: EASY/MEDIUM/HARD

**Overview/Abstract**: Description in less than 400 characters (5 lines, 80 chars).

### Description

Detailed description of the project.

### Linkography

* [Link1](https://localhost/THIS_IS_CLEARLY_A_BROKEN_LINK): Link1 explanation
* [Link2](https://localhost/THIS_IS_CLEARLY_A_BROKEN_LINK): Link2 explanation

> * [Link3](https://localhost/THIS_IS_CLEARLY_A_BROKEN_LINK): Link3 explanation

Reviewer: what is the usecase for "linkography"?

Author: It's used in case you have links that should to be available at a glance as external reference for the reader (say the project homepage, docs, etc..).

> ## Idea Example
> 
> This is an example of a real idea that was submitted https://github.com/NixOS/GSoC2024/pull/3
> 
> ### Metadata
> 
> **ID**: [2024/0001](./ideas/2024/0001)
> 
> **Title**: [nixpkgs] nixpkgs library networking functions
> 
> **Author**: FREE
> 
> **Proposed Menthor**: [@Janik-Haag](https://github.com/Janik-Haag)
> 
> **Effort**: SMALL
> 
> **Difficulty**: MEDIUM
> 
> **Overview/Abstract**: Add a set of built-in networking functions to
> `nixpkgs.lib` to improve contributors UX.
> 
> ### Description
> 
> Many things in NixOS use the Internet Protocol and other network related
> technologies, thus a lot of modules and service have to deal with them.
> Currently this is mostly done just passing strings as arguments, but it would
> be a lot more convenient to have things like `lib.types.ipv6`, `lib.types.ipv4`
> and other functions to for example convert between cidr notation and the bit
> representation of netmasks. This would not just be useful for NixOS it self but
> also for its users since a lot of folks use it to power network appliances and
> build infrastructure on top of it.
> 
> **Skills required/preferred**:
> Some prior networking knowledge would be useful but not required.
> Prior knowledge of the nix programming language is not required.
> Anyone new-comer wanting to pick this up should easily be able to complete a
> nix tutorial like [a tour of Nix](https://nixcloud.io/tour/?id=introduction/nix)
> in one or two afternoons and be good to go.
> 
> **Prior efforts**:
> There was some prior efforts in
> [nixpkgs#258250](https://github.com/NixOS/nixpkgs/pull/258250) but it is stuck
> for a few months now and also only adds function for legacy ip, it's also quite
> bare bones and we might want some more advanced things like a function to check
> if a ip matches a given subnet.
> 
> ### Linkography
> 
> * [A tour of Nix](https://nixcloud.io/tour/?id=introduction/nix)
> * [nixpkgs#258250](https://github.com/NixOS/nixpkgs/pull/258250)

Reviewer: ```suggestion
```
We should just link to a idea list from the year before

Author: yes but I don't think we have one now... or I couldn't find it.

I think we should links theirs for now ([PR#3]) and then fro next year on, link one from the year before.

Also, if we're going to reuse this repo for all years, it should probably be renamed (`GSoC2024` -> `GSoC` or `NixGSoC`)
